test/unit/xlog.c

```diff
@@ -0,0 +1,183 @@
+ * is roughly the same due to the randomness of the data.
+ */
+static void
+write_1k(struct xlog *xlog)
+{
+	/* Each uint64_t is encoded as 9 bytes + the array header. */
+	const size_t num_elements = 114;
+	const size_t data_size = num_elements * 9 + 3;
+
+	char data[data_size];
+	char *data_end = mp_encode_array(data, num_elements);
+	for (size_t i = 0; i < num_elements; i++)
+		data_end = mp_encode_uint(data_end, xoshiro_random());
+
+	if ((size_t)(data_end - data) != data_size)
+		panic("MsgPack encoding failed");
+	write_tuple(xlog, data, data_size);
+}
+
+/**
+ * Test that the size of the read buffer dynamically increased while reading a
+ * large file, and shrunk when xlog is written/read by small portions of data.
+ */
+static void
+test_dynamic_sized_ibuf(void)
+{
+	plan(4);
+	struct xlog xlog;
+	char dirname[] = "./xlog.XXXXXX";
+	create_xlog(&xlog, dirname);
+
+	/* Write about 20 MB of data to the xlog. */
+	for (int i = 0; i < 20 * 1024; i++)
+		write_1k(&xlog);
+	if (xlog_flush(&xlog) < 0)
+		panic("xlog_flush() failed");
+
+	struct xlog_cursor cursor;
+	if (xlog_cursor_open(&cursor, xlog.filename) < 0)
+		panic("xlog_cursor_open() failed");
+
+	/*
+	 * Read the whole xlog and check that the size of the buffer reaches
+	 * maximum value while reading, it will decrease when reading near the
+	 * end of the file, so keep it on each iteration.
+	 */
+	int rc;
+	int64_t prev_lsn = 0;
+	struct xrow_header row;
+	size_t read_ahead_max = 0;
+	size_t ibuf_used_max = 0;
+	while ((rc = xlog_cursor_next(&cursor, &row, false)) == 0) {
+		if (row.lsn != prev_lsn + 1)
+			panic("xlog is corrupted");
+		prev_lsn = row.lsn;
+
+		read_ahead_max = MAX(cursor.read_ahead, read_ahead_max);
+		ibuf_used_max = MAX(ibuf_used_max, ibuf_used(&cursor.rbuf));
+	}
+
+	is(read_ahead_max, XLOG_READ_AHEAD_MAX,
+	   "read_ahead increased to %d", XLOG_READ_AHEAD_MAX);
+	ok(ibuf_used_max >= XLOG_READ_AHEAD_MAX,
+	   "ibuf size increased to at least %d", XLOG_READ_AHEAD_MAX);
+
+	/*
+	 * Do a couple of 1 KB writes/reads to shrink the read buffer to the
+	 * minimal capacity.
+	 */
+	for (int i = 0; i < 10; i++) {
+		write_1k(&xlog);
+		if (xlog_flush(&xlog) < 0)
+			panic("xlog_flush() failed");
+
+		while ((rc = xlog_cursor_next(&cursor, &row, false)) == 0) {
+			if (row.lsn != prev_lsn + 1)
+				panic("xlog is corrupted");
+			prev_lsn = row.lsn;
+		}
+	}
+	size_t capacity = ibuf_capacity(&cursor.rbuf);
+	size_t start_capacity = cursor.rbuf.start_capacity;
+
+	is(cursor.read_ahead, XLOG_READ_AHEAD_MIN,
+	   "read_ahead decreased to %d", XLOG_READ_AHEAD_MIN);
+	ok(capacity >= start_capacity && capacity < 2 * start_capacity,
+	   "ibuf capacity decreased to [%zu..%zu)",
+	   start_capacity, 2 * start_capacity);
+
+	xlog_cursor_close(&cursor, false);
+	if (xlog_close(&xlog, false) < 0)
+		panic("xlog_close() failed");
+
+	char rmdir_cmd[sizeof("rm -rf ") + sizeof(dirname) + 1] = "rm -rf ";
+	strlcat(rmdir_cmd, dirname, sizeof(rmdir_cmd));
```

Comment:
`rmdir_cmd` is never used. Anyway, `rm -rf` looks dangerous. Let's explicitly remove only those files that we created with `unlink(2)` and `rmdir(2)`.

Reply:
Done